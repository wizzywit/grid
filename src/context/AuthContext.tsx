import React from "react";

interface DataArray {
  color: string;
  shape: string;
}

export type Filter = (
  c: Array<string> | ((c: Array<string>) => Array<string>)
) => void;

type PerformFilterType = (
  shape: string,
  shapesFilter: Array<string> | undefined,
  filter: Filter | undefined,
  type: string
) => void;

export type AuthContextType = {
  login?: () => void;
  logout?: () => void;
  data?: Array<DataArray>;
  shapesFilter?: Array<string>;
  setShapesFilter?: Filter;
  colorsFilter?: Array<string>;
  setColorsFilter?: Filter;
  isLoggedIn?: boolean;
  performSelect?: PerformFilterType;
  colors?: Array<string>;
  shapes?: Array<string>;
};
const AuthContext = React.createContext<AuthContextType>({});

export default AuthContext;
