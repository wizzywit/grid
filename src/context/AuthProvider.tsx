import AuthContext from "./AuthContext";
import { useState, useEffect, FC } from "react";
import axios from "axios";

type Props = {};

const AuthProvider: FC<Props> = ({ children }) => {
  const login = async () => {
    try {
      const response = await attemptLogin;
      return response;
    } catch (error) {
      console.log(error);
      throw new Error(error);
    }
  };

  const attemptLogin = new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({ status: 200, message: "successfully logged in" });
    });
  });
  const logout = () => {
    localStorage.removeItem("user");
    window.location.reload();
  };

  interface IUser {
    username: string;
  }

  const getCurrentUser = (): IUser | null => {
    var userStr = localStorage.getItem("user");
    if (userStr) {
      return JSON.parse(userStr || "");
    } else {
      return null;
    }
  };
  interface IResponse {
    items?: Array<{ color: string; shape: string }>;
  }
  const [response, setResponse] = useState<IResponse>({});
  const [data, setData] = useState<
    Array<{ color: string; shape: string }> | undefined
  >([]);
  const [shapesFilter, setShapesFilter] = useState([
    "oval",
    "traingle",
    "round",
    "square",
    "rectangle",
  ]);
  const [colorsFilter, setColorsFilter] = useState([
    "red",
    "blue",
    "green",
    "yellow",
    "cyan",
    "grey",
  ]);
  const getData = () => {
    axios
      .get("data.json", {
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
        },
      })
      .then((resp) => {
        console.log(resp);
        setResponse(resp.data);
        setData(resp.data.items);
      });
  };

  const filterResult = () => {
    const filteredArr = response.items?.filter((item) => {
      console.log(
        colorsFilter.includes(item.color),
        " ",
        shapesFilter.includes(item.shape)
      );

      return (
        colorsFilter.includes(item.color) && shapesFilter.includes(item.shape)
      );
    });
    setData(filteredArr);
  };

  useEffect(() => {
    getData();
  }, []);

  console.log(data);

  useEffect(() => {
    filterResult();
  }, [colorsFilter, shapesFilter]);

  const value = {
    login,
    logout,
    getCurrentUser,
    data,
    setShapesFilter,
    setColorsFilter,
    shapesFilter,
    colorsFilter,
  };
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export default AuthProvider;
