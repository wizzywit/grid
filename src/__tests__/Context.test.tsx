import React from "react";
import { render, screen } from "@testing-library/react";
import { AuthContextType } from "../context/AuthContext";

describe("Context", () => {
  test("should show value from provider", async () => {
    interface AType extends AuthContextType {
      test?: string;
    }
    const AuthContext = React.createContext<AType>({});
    const { container } = render(
      <AuthContext.Provider value={{ test: "testing" }}>
        <AuthContext.Consumer>
          {(value) => <span>Received: {value.test}</span>}
        </AuthContext.Consumer>
      </AuthContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent("Received: testing");
    expect(screen.getByText(/^Received:/).textContent).toBe(
      "Received: testing"
    );
  });
});
