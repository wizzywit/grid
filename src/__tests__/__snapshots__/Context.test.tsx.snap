// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context should show value from provider 1`] = `
<div>
  <span>
    Received: 
    testing
  </span>
</div>
`;
